### PR TITLE
[SPARK-27174][SQL] Add support for casting integer types to binary

### PR DESCRIPTION
**Pull request by s1ck**
Co-authored-by: Philip Stutz <philip.stutz@gmail.com>

## What changes were proposed in this pull request?

This PR adds support for casting

* `ByteType`
* `ShortType`
* `IntegerType`
* `LongType`

to `BinaryType`.

## How was this patch tested?

We added unit tests for casting instances of the above types. For validation, we used Javas `DataOutputStream` to compare the resulting byte array with the result of `Cast`.

We state that the contribution is our original work and that we license the work to the project under the project’s open source license.

@cloud-fan we'd appreciate a review if you find the time, thx

### CODE REVIEW

**Comment by maropu** [comment]
We need to support this cast?  Is it bad to just implement a built-in function like `to_binary`?

**Comment by cloud-fan** [comment]
CAST is a pretty standard SQL function, we should make sure our behavior align with the SQL spec and mainstream SQL systems.

**Comment by s1ck** [comment]
We stumbled across the missing functionality while doing prototyping work for https://issues.apache.org/jira/browse/SPARK-25994
Initially, we experimented with a more generalized built-in function (`serialize`/`to_binary`) that can convert several other types to binary arrays. 

During the experiment, we realized that such a general function is not needed, as we are not encoding the representation (i.e. using techniques such as var-length encoding) but merely interpret an integer as an array of bytes. That seemed like a perfect match for `Cast`.

In addition, a dedicated function would "promise" conversions over all kinds of values, while this PR just adds those that have an unambiguous binary representation.

**Comment by cloud-fan** [comment]
It may be convenient for some use cases, but it may also confuse others if only Spark has this behavior. Can you check with other databases?

**Comment by s1ck** [comment]
[DB2](https://www.ibm.com/support/knowledgecenter/en/SSEPEK_10.0.0/sqlref/src/tpc/db2z_castingbetweendatatypes.html#db2z_castingbetweendatatypes) does not support casting integer types to binary types.

[SQL Server](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-2017#implicit-conversions) supports even *implicit* conversions from integer types to binary types.

[Oracle](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/sqlrf/CAST.html#GUID-5A70235E-1209-4281-8521-B94497AAEF75) supports explicit casts from integer types to binary types.

We couldn't find a clear documentation of casting to binary in Postgres, MySQL and MariaDB to determine if this would work.
 
Do you suggest adding a custom expression exposed by a function (as proposed by @maropu)?

**Comment by cloud-fan** [comment]
what's the current behavior in Spark when casting int to binary? fail or return null?

**Comment by s1ck** [comment]
It throws
```
cannot resolve 'CAST(`a` AS BINARY)' due to data type mismatch: cannot cast int to binary;;
```

**Comment by cloud-fan** [comment]
Then I'm ok with supporting it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -1382,3 +1392,44 @@ case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[Str
   extends UnaryExpression with Unevaluable {
   override lazy val resolved = false
 }
+
+object ToBinary {
+
+  val objectName: String = ToBinary.getClass.getName.stripSuffix("$")
+
+  def cast(l: Long): Array[Byte] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: does JDK provide this functionality?

Comment (s1ck) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Java also provides the following way to convert `Long` to `byte[]`:

```
ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
buffer.putLong(x);
return buffer.array();
```

If we don't share the `ByteBuffer`, this is much slower than the custom conversion. We did a JMH benchmark which compares our custom conversion and the `ByteBuffer` approach:

```
import java.nio.ByteBuffer

import org.openjdk.jmh.annotations._

import scala.util.Random

object Impls {

  @inline final def customCast(l: Long): Array[Byte] = {
    val result = new Array[Byte](8)
    result(0) = (l >>> 56 & 0xFF).toByte
    result(1) = (l >>> 48 & 0xFF).toByte
    result(2) = (l >>> 40 & 0xFF).toByte
    result(3) = (l >>> 32 & 0xFF).toByte
    result(4) = (l >>> 24 & 0xFF).toByte
    result(5) = (l >>> 16 & 0xFF).toByte
    result(6) = (l >>> 8 & 0xFF).toByte
    result(7) = (l & 0xFF).toByte
    result
  }

  @inline final def byteBufferCast(l: Long): Array[Byte] = {
    val buffer = ByteBuffer.allocate(8)
    buffer.putLong(l)
    buffer.array
  }

}

@State(Scope.Benchmark)
@Fork(value = 3)
class LongSerializationBenchmark {

  var longsToConvert: Array[Long] = _

  @Setup
  def setUp(): Unit = {
    val numConversions = 1000
    longsToConvert = Array.fill(numConversions)(Random.nextLong)
  }

  @Benchmark
  def customCast(): Int = {
    val l = longsToConvert.length
    var check = 0
    var i = 0
    while (i < l) {
      check += Impls.customCast(longsToConvert(i)).length
      i += 1
    }
    check
  }

  @Benchmark
  def byteBufferCast(): Int = {
    val l = longsToConvert.length
    var check = 0
    var i = 0
    while (i < l) {
      check += Impls.byteBufferCast(longsToConvert(i)).length
      i += 1
    }
    check
  }

}
```

The results suggests that we should go for the custom implementation:

```
Benchmark                                   Mode  Cnt          Score         Error  Units
LongSerializationBenchmark.byteBufferCast  thrpt   15     165426.898 ±    3412.197  ops/s
LongSerializationBenchmark.customCast      thrpt   15  310527608.212 ± 6216611.765  ops/s
```

Is there a way to reuse the `ByteBuffer` across multiple conversions?

**Comment by maropu** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -44,6 +44,10 @@ object Cast {
     case (_, StringType) => true
 
     case (StringType, BinaryType) => true
+    case (ByteType, BinaryType) => true
+    case (ShortType, BinaryType) => true
+    case (IntegerType, BinaryType) => true
+    case (LongType, BinaryType) => true
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: `    case (_: IntegralType, BinaryType) => true`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 3]
```diff
@@ -836,6 +837,61 @@ class CastSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+    checkBinaryCast(0L)
+    checkBinaryCast(1L)
+    checkBinaryCast(-1L)
+    checkBinaryCast(Long.MaxValue)
+    checkBinaryCast(Long.MinValue)
+
+    dos.close()
+    bos.close()
+  }
+
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: I think we'd be better to add new tests in cast.sql: https://github.com/apache/spark/blob/cad475dcc9376557f882859856286e858002389a/sql/core/src/test/resources/sql-tests/inputs/cast.sql#L45

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: We can reuse object references and please see: https://github.com/apache/spark/blob/c324e1da9d0519fc13220202136307d4547b7426/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L854

**Comment by kiszk** [comment]
@s1ck Just a confirmation. This cast always assumes big-endian order for each type. Is it OK both for little-endian and little-endian?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 3 above)
Comment (s1ck) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: I added acceptance tests for integral values. To validate the output, I created a hex string. The pretty format for binary values seems to be their corresponding ASCII values.

**Comment by s1ck** [comment]
@kiszk Yes, the conversion uses big-endian. I think that's ok since Java always uses network order / big-endian regardless of the platform.

**Comment by s1ck** [comment]
@maropu I'm having some trouble fixing the expectation in this test: https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala#L715-L722 
It seems to encode the value as `KryoData` correctly. I can even `.show` the DataFrame, however, when collecting the rows, I get a `null` row. Could you please give me a hint where to look?

**Comment by maropu** [comment]
Have you checked https://issues.apache.org/jira/browse/SPARK-15910?
I think you need to change that test cuz this pr will support the `int -> binary` cast. How about using `double -> binary` in that test?

**Comment by s1ck** [comment]
Changed it to `double -> binary`. I was (and still am) confused that in that test `int -> binary` works for `df.show()` but not for `df.collect()`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 4]
```diff
@@ -1382,3 +1389,44 @@ case class UpCast(child: Expression, dataType: DataType, walkedTypePath: Seq[Str
   extends UnaryExpression with Unevaluable {
   override lazy val resolved = false
 }
+
+object ToBinary {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: can you move it to an individual file?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: and we need to write unit tests for `ToBinary`. This is the downside of using a custom implementation.

Comment (s1ck) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Do you suggest trying the approach @maropu proposed, i.e. sharing the `ByteBuffer`? I'm a bit skeptical about the performance implications as you need to reset the buffer etc.

Happy to move the current implementation to a separate file and write unit tests, though.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: The `ToBinary` doesn't reuse the byte[] as well, it's unfair to reuse ByteBuffer.

Comment (s1ck) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I moved the conversion logic into `NumberConverter`, an already existing utility object and added tests to the `NumberConverterSuite`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 3 above)
Comment (s1ck) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: I removed the tests from `CastSuite`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 5]
```diff
@@ -906,9 +911,13 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     }
   }
 
+  private val objectName: String = NumberConverter.getClass.getName.stripSuffix("$")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: this is not worth to be a `val`, can we inline it?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala [hunk 6]
```diff
@@ -37,4 +39,47 @@ class NumberConverterSuite extends SparkFunSuite {
...
+    checkToBinary(0L)
+    checkToBinary(1L)
+    checkToBinary(-1L)
+    checkToBinary(Long.MaxValue)
+    checkToBinary(Long.MinValue)
+  }
+
+  def checkToBinary[T](in: T): Unit = in match {
+    case b: Byte =>
+      assert(NumberConverter.toBinary(b) === ByteBuffer.allocate(1).put(b).array())
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala]: can we specify the byte order of `ByteBuffer` during test?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala [hunk 7]
```diff
@@ -37,4 +40,47 @@ class NumberConverterSuite extends SparkFunSuite {
...
+    checkToBinary(0L)
+    checkToBinary(1L)
+    checkToBinary(-1L)
+    checkToBinary(Long.MaxValue)
+    checkToBinary(Long.MinValue)
+  }
+
+  def checkToBinary[T](in: T): Unit = in match {
+    case b: Byte =>
+      assert(toBinary(b) === ByteBuffer.allocate(1).order(BIG_ENDIAN).put(b).array())
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala]: one question: Does the SQL standard say something about the endianness when casting integral to binary? What do other RDBMS do?

**Comment by shaneknapp** [comment]
test this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 8]
```diff
@@ -908,7 +913,11 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
 
   private[this] def castToBinaryCode(from: DataType): CastFunction = from match {
     case StringType =>
-      (c, evPrim, evNull) => code"$evPrim = $c.getBytes();"
+      (c, evPrim, evNull) =>
+        code"$evPrim = $c.getBytes();"
+    case ByteType | ShortType | IntegerType | LongType =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: `    case _: NumericType =>`

Comment (s1ck) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: changed it to `IntegralType`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala (hunk 7 above)
Comment (s1ck) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/NumberConverterSuite.scala]: As mentioned above, casting to binary is not specified within the SQL standard. From the vendor docs, I couldn't figure out how it's being handled. Imho, going big endian here (Java and network standard) is ok for an unspecified feature.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 8 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: oh, I missed...

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**